ring_doorbell/__init__.py

```diff
@@ -37,164 +22,19 @@ class Ring(object):
-
-                if 'auth' in self.cache:
-                    self.auth = self.cache['auth']
-
-                # test if token from cache_file is still valid and functional
-                # if not, it should continue to get a new auth token
-                url = API_URI + DEVICES_ENDPOINT
-                req = self.query(url, raw=True)
-                if req and req.status_code == 200:
-                    self._authenticate(session=req)
-                else:
-                    self._authenticate()
-        else:
-            # first time executing, so we have to create a cache file
-            self._authenticate()
-
-    def _get_oauth_token(self):
-        """Return Oauth Bearer token."""
-        # this token should be cached / saved for later
-        oauth = Auth(self.auth)
-
-        if not self.auth:
-            self.auth = oauth.fetch_token(
-                self.username,
-                self.password,
-                self.auth_callback)
-        else:
-            if self.last_refresh:
-                expires_in = timedelta(seconds=self.auth['expires_in'])
-                refresh_at = self.last_refresh + expires_in
-
-                if self.debug:
-                    _LOGGER.debug("response from get oauth token %s",
-                                  str(self.auth))
-
-            if not self.last_refresh or (dt.now() >= refresh_at):
-                self.auth = oauth.refresh_tokens()
-                self.last_refresh = dt.now()
-            else:
-                if self.debug:
-                    _LOGGER.debug("Reusing oauth token %s", str(self.auth))
-
-        return self.auth['access_token']
-
-    def _authenticate(self, attempts=RETRY_TOKEN, session=None, wait=1.0):
-        """Authenticate user against Ring API."""
-        url = API_URI + NEW_SESSION_ENDPOINT
-        loop = 0
-        # make a copy as we're mutating headers in the loop below
-        # which would cause issues with _get_oauth_token()
-        # which expects a non mutated HEADERS copy
-        modified_headers = HEADERS.copy()
-        while loop <= attempts:
-            modified_headers['Authorization'] = \
-                'Bearer {}'.format(self._get_oauth_token())
-            loop += 1
-
-            try:
-                if session is None:
-                    req = self.session.post((url),
-                                            data=POST_DATA,
-                                            headers=modified_headers,
-                                            timeout=self._timeout)
-                else:
-                    req = session
-            except requests.exceptions.RequestException as err_msg:
-                _LOGGER.error("Error!! %s", err_msg)
-                raise
-
-            if not req:
-                time.sleep(wait)  # add a pause or you'll get rate limited
-                continue
+        self._oauth = Auth()
+        self._oauth.fetch_token(username, password, auth_callback)
```

Comment:
We shouldn't always require username and password. 

I would suggest that we create a standalone method to convert user/pass to token. This will create the first cache. This cache can then be passed to the Auth object and be kept up to date.

The way it now works in Home Assistant is that it requests user/pass once during setup, after that it just works with tokens. There is no reason to pass user/pass in all the time.